lib/compiler/passes/generate-js.js

```diff
@@ -11,11 +11,78 @@ function generateJS(ast, options) {
   function indent2(code)  { return code.replace(/^(.+)$/gm, "  $1");         }
   function indent10(code) { return code.replace(/^(.+)$/gm, "          $1"); }
 
+  function l(i) { return "peg$c" + i; } // |literals[i]| of the abstract machine
+  function r(i) { return "peg$r" + i; } // |classes[i]| of the abstract machine
+  function e(i) { return "peg$e" + i; } // |expectations[i]| of the abstract machine
+  function f(i) { return "peg$f" + i; } // |actions[i]| of the abstract machine
+
   function generateTables() {
+    function buildLiteral(literal) {
+      return "\"" + js.stringEscape(literal) + "\"";
+    }
+
+    function buildRegexp(cls) {
+      return "/^["
+            + (cls.inverted ? "^" : "")
+            + cls.value.map(part =>
+                Array.isArray(part)
+                  ? js.regexpClassEscape(part[0])
+                    + "-"
+                    + js.regexpClassEscape(part[1])
+                  : js.regexpClassEscape(part)
+              ).join("")
+            + "]/" + (cls.ignoreCase ? "i" : "");
+    }
+
+    function buildExpectation(e) {
+      switch (e.type) {
+        case "rule": {
+          return "peg$otherExpectation(\"" + js.stringEscape(e.value) + "\")";
+        }
+        case "literal": {
+          return "peg$literalExpectation(\""
+                  + js.stringEscape(e.value)
+                  + "\", "
+                  + e.ignoreCase
+                  + ")";
+        }
+        case "class": {
+          const parts = e.value.map(part =>
+            Array.isArray(part)
+              ? "[\"" + js.stringEscape(part[0]) + "\", \"" + js.stringEscape(part[1]) + "\"]"
+              : "\""  + js.stringEscape(part) + "\""
+          ).join(", ");
+
+          return "peg$classExpectation(["
+                  + parts + "], "
+                  + e.inverted + ", "
+                  + e.ignoreCase
+                  + ")";
+        }
+        case "any": return "peg$anyExpectation()";
+        default: throw new Error("Unknown expectation type (" + JSON.stringify(e) + ")");
+      }
+    }
+
+    function buildFunc(a) {
+      return "function(" + a.params.join(", ") + ") {"
+             + a.body
+             + "}";
+    }
+
     if (options.optimize === "size") {
```

Comment:
i wish we had gotten rid of the size optimization (#92) before this landed.

Reply:
it seems almost certain the perf people could really work with these chains of concats

Reply:
As you wish. I can prepare PR for deletion of that mode and then rebase on another and vice versa, depending on which of their would merged first

Reply:
If you're willing to do that, I think it would be really great.  I'll probably land GrammarError next then, while you work on that?

Reply:
OK